diffrax/_step_size_controller/jump_step_wrapper.py

```diff
@@ -0,0 +1,410 @@
+        jump_ts=None,
+        rejected_step_buffer_len=0,
+        _callback_on_reject=None,
+    ):
+        r"""
+        **Arguments**:
+
+        - `controller`: The controller to wrap.
+            Can be any diffrax.AbstractStepSizeController.
+        - `step_ts`: Denotes extra times that must be stepped to.
+        - `jump_ts`: Denotes extra times that must be stepped to, and at which the
+            vector field has a known discontinuity. (This is used to force FSAL solvers
+            so re-evaluate the vector field.)
+        - `rejected_step_buffer_len`: The length of the buffer storing rejected steps.
+            If this is > 0, then the controller will revisit rejected steps. This is
+            useful for SDEs, where the solution is guaranteed to be correct only if the
+            SDE is evaluated at all times at which the Brownian motion (BM) is
+            evaluated. Since the BM is also evaluated at rejected steps, we must later
+            evaluate the SDE at these times as well.
+        """
+        self.controller = controller
+        self.step_ts = _none_or_array(step_ts)
+        self.jump_ts = _none_or_array(jump_ts)
+        self.rejected_step_buffer_len = rejected_step_buffer_len
+        self.callback_on_reject = _callback_on_reject
+        self.__check_init__()
+
+    def __check_init__(self):
+        if self.jump_ts is not None and not jnp.issubdtype(
+            self.jump_ts.dtype, jnp.inexact
+        ):
+            raise ValueError(
+                f"jump_ts must be floating point, not {self.jump_ts.dtype}"
+            )
+
+    def wrap(self, direction: IntScalarLike):
+        step_ts = None if self.step_ts is None else self.step_ts * direction
+        jump_ts = None if self.jump_ts is None else self.jump_ts * direction
+        return eqx.tree_at(
+            lambda s: (s.step_ts, s.jump_ts),
+            self,
+            (step_ts, jump_ts),
+            is_leaf=lambda x: x is None,
+        )
+
+    def init(
+        self,
+        terms: PyTree[AbstractTerm],
+        t0: RealScalarLike,
+        t1: RealScalarLike,
+        y0: Y,
+        dt0: _Dt0,
+        args: Args,
+        func: Callable[[PyTree[AbstractTerm], RealScalarLike, Y, Args], VF],
+        error_order: Optional[RealScalarLike],
+    ) -> tuple[RealScalarLike, _JumpStepState[_ControllerState]]:
+        t1, inner_state = self.controller.init(
+            terms, t0, t1, y0, dt0, args, func, error_order
+        )
+        dt_proposal = t1 - t0
+        tdtype = jnp.result_type(t0, t1)
+
+        if self.step_ts is not None:
+            # Upcast step_ts to the same dtype as t0, t1
+            step_ts = upcast_or_raise(
+                self.step_ts,
+                jnp.zeros((), tdtype),
+                "`PIDController.step_ts`",
+                "time (the result type of `t0`, `t1`, `dt0`, `SaveAt(ts=...)` etc.)",
+            )
+        else:
+            step_ts = None
+
+        if self.jump_ts is not None:
```

Comment:
Nit: remove the `not` and flip the branches. We can save four characters ;)

(By which I mean, I find the logic easier to follow.)